cookbooks/retention.md

```diff
@@ -0,0 +1,211 @@
+
+First let's map 1, 2, ..., N week retention the the amount of days elapsed after the anchor point:
+
+```python
+PERIODS = {}
+N_WEEKS = 6
+for i in range(1, N_WEEKS + 1):
+    PERIODS[i] = {
+        'start': i * 7,
+        'end': i * 7 + 6
+    }  
+```
+Which gives us
+
+```python
+{1: {'end': 13, 'start': 7},
+ 2: {'end': 20, 'start': 14},
+ 3: {'end': 27, 'start': 21},
+ 4: {'end': 34, 'start': 28},
+ 5: {'end': 41, 'start': 35},
+ 6: {'end': 48, 'start': 42}}
+
+```
+
+Next, let's define some helper functions: 
+
+```python
+import datetime as dt
+import pandas as pd
+import pyspark.sql.types as st
+import pyspark.sql.functions as F
+
+udf = F.udf
+
+def date_diff(d1, d2, fmt='%Y%m%d'):
+    """
+    Returns days elapsed from d2 to d1 as an integer
+    
+    Params:
+    d1 (str)
+    d2 (str)
+    fmt (str): format of d1 and d2 (must be the same)
+    
+    >>> date_diff('20170205', '20170201')
+    4
+    
+    >>> date_diff('20170201', '20170205)
+    -4
+    """
+    try:
+        return (pd.to_datetime(d1, format=fmt) - 
+                pd.to_datetime(d2, format=fmt)).days
+    except:
+        return None
+    
+
+@udf(returnType=st.IntegerType())
+def get_period(anchor, submission_date_s3):
+    """
+    Given an anchor and a submission_date_s3,
+    returns what period a ping belongs to. This 
+    is a spark UDF.
+    
+    Params:
+    anchor (col): anchor date
+    submission_date_s3 (col): a ping's submission_date to s3
+    
+    Global:
+    PERIODS (dict): defined globally based on n-week method
+    
+    Returns an integer indicating the retention period
+    """
+    if anchor is not None:
+        diff = date_diff(submission_date_s3, anchor)
+        if diff >= 7: # exclude first 7 days
+            for period in sorted(PERIODS):
+                if diff <= PERIODS[period]['end']:
+                    return period
+
+@udf(returnType=st.StringType())
+def from_unixtime_handler(ut):
+    """
+    Converts unix time (in days) to a string in %Y%m%d format.
+    This is spark UDF.
+    
+    Params:
+    ut (int): unix time in days
+    
+    Returns a date as a string if it is parsable by datetime, otherwise None
+    """
+    if ut is not None:
+        try:
+            return (dt.datetime.fromtimestamp(ut * 24 * 60 * 60).strftime("%Y%m%d"))
+        except:
+            return None
+        
+```
+
+Now we can load in a subset of `main_summary` and construct the necessary fields for retention calculations:
+
+```python
+ms = spark.sql("""
```

Comment:
Nit: this SQL statement should follow the [style guide](https://docs.telemetry.mozilla.org/concepts/sql_style.html). In particular, please:

* Capitalize reserved words like `WHERE`, `SELECT`, and `AND` ([link](https://docs.telemetry.mozilla.org/concepts/sql_style.html#reserved-words))
* Add a newline after root keywords like `SELECT` and `WHERE` ([link](https://docs.telemetry.mozilla.org/concepts/sql_style.html#code-blocks))
* Indent lines 126-130 one additional level to clarify the `WHERE` statement block